### PR TITLE
[Core] Reduce CBlockIndex RAM usage

### PR DESCRIPTION
**Pull request by random-zebra**
This completes the work started in #1338 and bumps the version to `4.0.99.2`.

It gives a dramatic improvement in memory usage.
On mainnet synced up to block 2281117, it takes **~926 Mb** in total (a 59% reduction over the 2.261 Gb of `v4.0.99.1`, and 64% overall reduction over the 2.570 Gb of `v4.0.2`).

Changes:
`mapZerocoinSupply` and `moneySupply` are removed from the block index serialization and replaced with a single cached map/value updated in `ConnectBlock`/`DisconnectBlock` and databased independently.

To avoid reindex on upgrade, a utility class `CLegacyBlockIndex` is introduced, in order to marshall old (removed) block index members directly from disk.

### CODE REVIEW

File: src/zpivchain.cpp [hunk 1]
```diff
@@ -392,3 +394,15 @@ std::list<libzerocoin::CoinDenomination> ZerocoinSpendListFromBlock(const CBlock
     return vSpends;
 }
 
+int64_t GetZerocoinSupply()
+{
+    if (mapZerocoinSupply.empty())
```
Comment (furszy) [inline src/zpivchain.cpp:408]: Something that have in mind, as now `mapZerocoinSupply` works under an specific context (tip), we should add an `AssertLockHeld(cs_main);` here and  double check that every line viewing/modifying `mapZerocoinSupply` is previously locking `cs_main`.

File: src/chain.h [hunk 2]
```diff
@@ -325,7 +322,11 @@ class CDiskBlockIndex : public CBlockIndex
             READWRITE(nBits);
             READWRITE(nNonce);
             if(this->nVersion > 3) {
-                READWRITE(mapZerocoinSupply);
+                // zc supply removed in 4.0.99.2
+                if (nSerVersion < DBI_SER_VERSION_NO_ZC) {
+                    std::map<libzerocoin::CoinDenomination, int64_t> mapZerocoinSupply;
+                    READWRITE(mapZerocoinSupply);
```
Comment (furszy) [inline src/chain.h]: is the write operation ever used here? could be just `READDATA`?

File: src/zpivchain.cpp (hunk 1 above)
Comment (random-zebra) [reply src/zpivchain.cpp:408]: sure

File: src/chain.h (hunk 2 above)
Comment (random-zebra) [reply src/chain.h]: wouldn't use basic operations outside serialize.h. We can simply wrap it inside `if (ser_action.ForRead())` as it is done elsewhere.

**Comment by random-zebra** [comment]
Updated as per feedback.

**Comment by Fuzzbawls** [comment]
So far, so good
ScreenShot_20200410042245

**Pull request merged**